internal/v4/api_test.go

```diff
@@ -681,36 +683,45 @@ func (s *APISuite) TestMetaPerm(c *gc.C) {
 		},
 	})
 
-	// Publish one of the revisions to development and check that the development ACLs
-	// have changed.
-	err = s.store.Publish(newResolvedURL("~charmers/precise/wordpress-23", 23), mongodoc.DevelopmentChannel)
+	// Publish one of the revisions to development, then PUT to meta/perm
+	// and check that the development ACLs have changed.
+	err := s.store.Publish(newResolvedURL("~charmers/precise/wordpress-23", 23), mongodoc.DevelopmentChannel)
 	c.Assert(err, gc.IsNil)
-	s.assertPut(c, "precise/wordpress-23/meta/perm/read", []string{"bob", "charlie"})
 
-	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
-		Handler: s.srv,
-		Do:      bakeryDo(nil),
-		URL:     storeURL("precise/wordpress-23/meta/perm"),
-		ExpectBody: params.PermResponse{
-			Read:  []string{"bob", "charlie"},
-			Write: []string{"charmers"},
-		},
+	s.doAsUser("bob", func() {
+		// Check that we aren't allowed to put to the newly published entity as bob.
+		s.assertPutUnauthorized(c, "~charmers/precise/wordpress/meta/perm/read?channel=development", []string{}, `unauthorized: access denied for user "bob"`)
 	})
 
-	// The other revisions should still see the old ACLs.
-	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
-		Handler: s.srv,
-		Do:      bakeryDo(nil),
-		URL:     storeURL("precise/wordpress-24/meta/perm"),
-		ExpectBody: params.PermResponse{
-			Read:  []string{"bob"},
-			Write: []string{"admin"},
-		},
+	s.doAsUser("charmers", func() {
+		s.discharge = dischargeForUser("charmers")
+		s.assertPut(c, "precise/wordpress-23/meta/perm/read", []string{"bob", "charlie"})
+		s.assertGetUnauthorized(c, "~charmers/precise/wordpress/meta/perm/read?channel=development", `unauthorized: access denied for user "charmers"`)
 	})
 
-	e, err = s.store.FindBaseEntity(charm.MustParseURL("precise/wordpress-23"), nil)
-	c.Assert(err, gc.IsNil)
-	c.Assert(e.ChannelACLs, gc.DeepEquals, map[mongodoc.Channel]mongodoc.ACL{
+	s.doAsUser("bob", func() {
+		httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+			Handler: s.srv,
+			Do:      bakeryDo(nil),
```

Comment:
There is something tricksy here having to know to use bakeryDo if you want you're doAsUser wrapper to work. I don't think I'm too keen on this pattern. I think it will lead to confusion for no real benefit.